prometheus/collector.go

```diff
@@ -0,0 +1,156 @@
+package prometheus
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	exporter "github.com/kawamuray/prometheus-kafka-consumer-group-exporter"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	partitionOffsetMetricsDesc = prometheus.NewDesc(
+		"kafka_broker_consumer_group_current_offset",
+		"Current consumed offset of a topic/partition",
+		[]string{"group_id", "consumer_address", "client_id", "topic", "partition"},
+		nil)
+	partitionLagMetricsDesc = prometheus.NewDesc(
+		"kafka_broker_consumer_group_offset_lag",
+		"Offset lag of a topic/partition",
+		[]string{"group_id", "consumer_address", "client_id", "topic", "partition"},
+		nil)
+)
+
+// PartitionInfoCollector is a Kafka prometheus.Collector. It uses a
+// exporter.PartitionInfoClient for the actual querying of Kafka.
+//
+// To speed up collection each consumer group is collected
+// concurrently.
+type PartitionInfoCollector struct {
+	groupListErrors     prometheus.Counter
+	groupDescribeErrors *prometheus.CounterVec
+
+	client exporter.PartitionInfoClient
+
+	ctx         context.Context
+	execTimeout time.Duration
+}
+
+// NewPartitionInfoCollector returns a prometheus.Collector that queries Kafka
+// using client. concurrency sets an upper limit on the number concurrent Kafka
+// concumer group queries running.
+func NewPartitionInfoCollector(client exporter.PartitionInfoClient, ctx context.Context, execTimeout time.Duration) *PartitionInfoCollector {
+	return &PartitionInfoCollector{
+		prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_list_errors",
+			Help: "Number of Kafka scraping errors.",
+		}),
+		prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_describe_errors",
+			Help: "Number of Kafka scraping errors.",
+		}, []string{"group"}),
+		client,
+		ctx,
+		execTimeout,
+	}
+}
+
+// Describe transmits all metric descriptions to c.
+func (p *PartitionInfoCollector) Describe(c chan<- *prometheus.Desc) {
+	c <- partitionOffsetMetricsDesc
+	c <- partitionLagMetricsDesc
+	p.groupListErrors.Describe(c)
+	p.groupDescribeErrors.Describe(c)
+}
+
+// Collect triggers an on-demand scraping from Kafka and transmits metrics into
+// c.
+func (p *PartitionInfoCollector) Collect(c chan<- prometheus.Metric) {
+	// Important that these are collected _after_ the Kafka collection below to
+	// correctly accomodate for the errors that happened during the scrape.
+	defer p.groupDescribeErrors.Collect(c)
+	defer p.groupListErrors.Collect(c)
+
+	ctx, cancel := context.WithTimeout(p.ctx, p.execTimeout)
+	groupnames, err := p.client.Groups(ctx)
+	cancel()
+	if err != nil {
+		log.Error("Could not list groups:", err)
+		p.groupListErrors.Inc()
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(groupnames))
+	for _, groupname := range groupnames {
+		go func(groupname string) {
```

Comment:
Good idea. We currently only have a single topic and will soon be adding a second one. I added a default max concurrency limit of 20 in ab8217ed7eedd28d39e511134328401295ad7b06. Do you think that's a sane limit?